spec/ruby/core/enumerable/grep_spec.rb

```diff
@@ -40,15 +40,12 @@ class EnumerableSpecGrep2; def ===(obj); /^ca/ =~ obj; end; end
     $~.should == nil
   end
 
-  it "sets $~ to the last match when given no block" do
-    "z" =~ /z/ # Reset $~
-    ["abc", "def"].grep(/b/).should == ["abc"]
-
-    # Set by the failed match of "def"
-    $~.should == nil
-
-    ["abc", "def"].grep(/e/)
-    $&.should == "e"
+  ruby_version_is "3.0.0" do
```

Comment:
Usually when changing behavior, we guard the existing spec with `ruby_version_is ...'3.0.0'` and also add a new spec for the Ruby 3 behavior.  Is there a reason for removing the spec in older versions?

Reply:
Looks like the spec was added a few days after I opened my issue by @eregon in 9cf4767f05487464b768f02461d6d0879d9887cb, which surprises me a little. In this case, I feel that this was never intended and should be considered a bug or at the very least unspecified behavior (which is why I removed it). There used to be `ruby_bug` to spec such behavior (when it was important), but there's a single use now. @eregon: what's the preferred way to handle this?

Reply:
https://github.com/ruby/spec/commit/9cf4767f05487464b768f02461d6d0879d9887cb was in 2019 and https://bugs.ruby-lang.org/issues/17030 was in 2020.
That's an undeserved and uncalled for implication.
FWIW I added the specs and fixed behavior in TruffleRuby in https://github.com/oracle/truffleruby/commit/7875a92c3f0fba33bec0ced66f25e4b48171a836, which I think I found while investigating issues related to $~ which is very complex to implement correctly and efficiently.

What @jeremyevans said, keep the existing spec and add a version guard, it is the behavior for existing versions, and changing it in < 3 would be a breaking change, the spec is there to test the existing behavior.

`ruby_bug` is basically just to prevent on CRuby segfaults, hangs, or determined as completely incorrect result after discussion on a ticket, and when the correct behavior is clearly known.

In general, I agree it rarely makes much sense to use `$~` for `grep/grep_v` without a block, which I already said in https://bugs.ruby-lang.org/issues/17030#note-5
(OTOH, all the other matching methods do set it, even without a block)

Reply:
> That's an undeserved and uncalled for implication.

Oh, sorry I confused the years 🤦‍♂️. I also didn't mean to imply anything (positive or negative), I'm sorry if it reads differently.

> keep the existing spec and add a version guard

Done!

> all the other matching methods do set it, even without a block

Which ones are these? Maybe we can optimize those too...

Reply:
@marcandre clarified with me, I misinterpreted what he said earlier.